### ISSUE TITLE
Feature request: colour grading

### ISSUE THREAD

**Issue by roborourke** [lovell/sharp#1441]
I'm interested in being able to apply 3D LUTs to images on the fly, eg. `.cube` files. I think `vips_maplut()` is the right function under the hood for this, perhaps I'll need to figure out how to generate LUT images from those files first though.

I've been trying to use some node libraries to do this, extracting the pixels as an ndarray, applying the LUT, then saving the modified pixels back to an image buffer but it's super slow and error prone (truncated image data etc).

If it's not something that makes sense for sharp no worries, hopefully I can find info on transforming an image as a `readableStream` assuming that's doable :)

**Comment by lovell** [comment]
Hello, you're correct that libvips' `maplut` operation is suitable for this - we'd need to expose it in sharp. Happy to accept/help with a PR for this if you're able.

**Comment by roborourke** [comment]
@lovell I've been tinkering with it for a while but I think I'm not quite understanding something. It looks like I can only pass a single image to `maplut()`, I'm trying with a hald image from https://filterhunt.com but as far as I understand it you need two hald images - one with the standard RGB values that is then mapped to the LUT hald image. Do you know if that's right?

I have some work in progress here: https://github.com/roborourke/sharp/tree/maplut

It compiles at least! Doesn't actually modify the output at all though.

**Comment by lovell** [comment]
Ah, sorry, I was thinking of 3x 1D LUTs, e.g. one for each channel of an RGB image. For `maplut` I guess a HALD image would need to be parsed/converted/interpolated into separate R, G and B tables first, if such a thing is possible, then combined into a 3-channel image.

It might be worth you asking this question over on the [libvips](https://github.com/libvips/libvips/) repo. Perhaps a 3D LUT could be converted to an ICC profile and applied that way?

**Comment by roborourke** [comment]
Yeah you can get them in icm format I think. I see your point re 3x 1D LUTs. Hopefully there might be some existing tools to convert between LUT formats :/

Will ask over on the libvips repo, thanks!

**Comment by lovell** [comment]
https://github.com/libvips/libvips/issues/1153

**Comment by roborourke** [comment]
Thanks for cross-linking! I should have done that. I've played a bit more using `vips_icc_transform()` but it still doesn't seem to modify the output so I'm definitely doing something wrong...

Work in progress: https://github.com/roborourke/sharp/tree/icc-transform

**Comment by lovell** [comment]
If you didn't see it, #1323 is the existing feature request relating to custom output profiles.

**Comment by roborourke** [comment]
Cheers, I guess I didn't know what I was looking for before :) So the intention in #1323 is to modify the output pixel values? Would it matter that it didn't have the sRGB colourspace then?

Anyway, if this use case is covered by that issue then we can close this one.

**Comment by lovell** [comment]
If you think you'll be able to achieve what you need using a colour profile transform (rather than a 3D LUT) then yes, the possible future enhancement discussed in #1323 should suffice.

I'm happy to keep this issue open as another possible future enhancement given others may require a LUT feature, albeit limited to per-channel manipulation.

**Comment by roborourke** [comment]
Commented on #1323 - the `icc-transform` branch on my fork is working for applying a custom ICC profile now but I'm not sure it's definitely working the same way a 3D LUT would. Given the capabilities of libvips right now it might be the closest I can get for the time being at least!

**Comment by lovell** [comment]
Note to self/implementer: we could provide some pre-canned LUTs e.g. Viridis - see https://github.com/lovell/sharp/issues/4287

**Comment by StellarCodeOrg** [comment]
Just providing some more illustration of the nice things that could be achieved with such a feature: 
https://blog.frost.kiwi/Davinci-Resolve-thermal-luts/

By the way some colorscale LUT files (.cube, i.e. basically text files with 3 columns) can be downloaded from the middle of the page (could be useful in case of pre-canned implementation in sharp).

And for more context, here's a useful video about colormapping (from the authors of the viridis palette): https://www.youtube.com/watch?v=xAoljeRJ3lU. 

The feature certainly deserves its place in a library like sharp! (it would open a wide range of scientific applications, from medical imagery to anything one can imagine).

**Comment by StellarCodeOrg** [comment]
Hi Lovell,

To help a potentiel helper/implementer get started on the topic, is it possible to lay out the 3 big steps necessary to get the work done? 

The "big picture" necessary tasks in order to expose libvips maplut in sharp (and maybe the 3 big substeps in each one)?

Maybe it will help us to progress on this issue.

**Comment by lovell** [comment]
The public API might look something like:
```ts
// PROPOSED API, NOT YET AVAILABLE
maplut(lookupTable: Array<number> | Array<Array<number>>)
```

...where `lookupTable` is an array of 256 integers, or an array containing 3 or 4 arrays, each of 256 integers.

The JS logic probably makes sense added alongside the [channel-based operations](https://github.com/lovell/sharp/blob/main/lib/channel.js).

The C++ logic will need to build an `VImage` from this data with 1, 3 or 4 channels and pass this as `lut` to libvips' `maplut()`.

https://www.libvips.org/API/current/libvips-histogram.html#vips-maplut
